src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -325,6 +325,23 @@ public List<ChainedHeader> PartialValidationSucceeded(ChainedHeader chainedHeade
                 return null;
             }
 
+            // Can happen in case of a race condition when peer 1 presented a block, we started partial validation, peer 1 disconnected,
+            // peer 2 connected, presented header and supplied a block and block puller pushed it so the block data is not null.
+            if ((chainedHeader.Previous.BlockValidationState != ValidationState.PartiallyValidated) &&
+                (chainedHeader.Previous.BlockValidationState != ValidationState.FullyValidated))
+            {
+                this.logger.LogTrace("(-)[PREV_BLOCK_NOT_VALIDATED]:null");
+                return null;
+            }
+
+            // Same scenario as above except for prev block was validated which triggered next partial validation to be started.
+            if ((chainedHeader.BlockValidationState == ValidationState.PartiallyValidated) || 
+                (chainedHeader.BlockValidationState == ValidationState.FullyValidated))
+            {
+                this.logger.LogTrace("(-)[ALREADY_VALIDATED]:null");
+                return null;
```

Comment:
In this case we should return an empty `headersToValidate` no?

Reply:
no, why making an additional allocation?